500-CNAB-dependencies.md

```diff
@@ -0,0 +1,100 @@
+---
+title: Dependencies
+weight: 500
+---
+
+# CNAB Dependencies 1.0
+Working Draft, April. 2019
+
+This specification describes how bundles can define dependencies on other
+bundles.
+
+This specification, the CNAB Dependencies specification, is not part of the CNAB Core specification. An implementation of CNAB Core MAY NOT implement this specification, yet still claim compliance with CNAB Core. A CNAB implementation MUST NOT claim to be CNAB Dependencies-compliant unless it meets this specification.
+
+# Overview
+
+One way for a bundle author to manage the complexity of a large bundle and reuse
+common logic is to create bundles for discrete components, such as managing
+a database as a service on a cloud platform, and then use that bundle as a 
+dependency.
+
+This specification defines [dependencies metadata](#dependencies-metadata) in the 
+bundle.json for specifying dependencies but does not dictate that the metadata is
+specifically used at a particular time, or how the dependency graph is resolved. 
+Each tool may choose to use the information differently, for example [porter](https://porter.sh) resolves dependencies when the bundle is built, but another tool could 
+use this information at runtime.
+
+There are two cases for how a bundle may need to depend upon another bundle:
+
+1. [Depend on a named bundle](#depend-on-a-named-bundle)
+1. [Depend on a bundle that provides a capability](#depend-on-a-bundle-that-provides-a-capability)
+
+## Depend on a named bundle
+
+The bundle depends on a specific named bundle that is known in advance.
+
+```json
+{
+    "name": "wordpress",
+    "dependencies": [
+        {
+            "requires":{
+                "bundle": "azure/mysql",
+                "version": {
+                    "range": "5.7.x",
+                    "prereleases": "true"
+                }
+            }
+        }
+    ]
+}
+```
+
+## Depend on a bundle that provides a capability
+
+This section is a placeholder and will be completed in a follow-up pull request.
+
+## Dependencies Metadata
+
+This specification introduces a `dependencies` array in the bundle.json
+that defines metadata necessary to specify a dependency.
+
+### requires
+
+The `requires` object defines the criteria for the dependent bundle:
+
+* `bundle`: A reference to a bundle in the format REGISTRY/NAME.
+* `version`: A set of criteria applied to the bundle version when selecting an
```

Comment:
I have a question, which may be dummy, about the versions and the reference to a bundle in the format `REGISTRY/NAME`:
Which version are we talking about? The one defined in the bundle.json, or the tag to reference the right bundle version in a registry `REGISTRY/NAME:TAG` ?

My point is:
- if we use the `version` field in the bundle, then the resolver will need to pull all the bundles, parse them, then resolve the best version. I guess this can become heavy with time and multiple versions.
- if we use the `tag`, the resolver has just to fetch all the tags to select the best version, but then it means we must tight the tag with the version field, as there is no way to guarantee that the tag is relevant with its version.

Does my question make sense? Am I missing something?

Reply:
My  intention was to use the latter, where the bundle author is expected to tag their bundles with the bundle version so that people using this dependency spec can depend upon this convention.

Do you anticipate a problem with that or is that a reasonable convention that people can follow?

Reply:
The registry conversation and how duffle pushes bundles to registries right now makes it likely that bundles will be tagged with a digest and not semver. What is your thinking around referring to a bundle name with its associated tag all in the `BUNDLE` field? Perhaps that misses the point since you'd have to create a new bundle for every dependency bump.

Reply:
> it likely that bundles will be tagged with a digest and not semver.

With `docker app` we push the bundles with a tag, and use `semver` by default (even if the user can choose any tag if she wants), and I guess most of the CNAB tools will do the same?

> Do you anticipate a problem with that or is that a reasonable convention that people can follow?

No that's a perfectly reasonable convention 👍 I don't anticipate a problem, just wanted to make sure we were talking about using `tags` for the dependency management. That said, maybe it should be added in the specifications? Or maybe is it purely an implementation detail?